batch/batch/worker/worker.py

```diff
@@ -1086,13 +1097,30 @@ async def run(self):
                         headers=self.headers)
                 log.info('deactivated')
         finally:
-            log.info('shutting down')
-            if site:
-                await site.stop()
-                log.info('stopped site')
-            if app_runner:
-                await app_runner.cleanup()
-                log.info('cleaned up app runner')
+            await self.cleanup()
+            self.shutdown()
+
+    async def kill(self, request):
+        body = await request.json()
+
+        expected_name = body['name']
+        if expected_name != NAME:
+            return web.HTTPForbidden()
+
+        try:
+            async with client_session(
+                    raise_for_status=True, timeout=aiohttp.ClientTimeout(total=60)) as session:
+                # Don't retry.  If it doesn't go through, the driver
+                # monitoring loops will recover.  If the driver is
+                # gone (e.g. testing a PR), this would go into an
+                # infinite loop and the instance won't be deleted.
+                await session.post(
+                    deploy_config.url('batch-driver', '/api/v1alpha/instances/deactivate'),
+                    headers=self.headers)
+            log.info('deactivated')
+        finally:
+            await self.cleanup()
+            self.shutdown()
```

Comment:
this feels a bit funky. The coroutine executing `run` will stay open until the idle time gets long. That while loop in `run` should have a way to indicate you want to stop. Then you don't need to duplicate the deactivate and shutdown code at all.

Reply:
Something like replacing the sleep with:
```
try:
    await asyncio.wait_for(stop_event.wait(), 15)
    break
except asyncio.TimeoutError:
    idle_duration = time_msecs() - self.last_updated
```